tests/regression/src/lib.rs

```diff
@@ -55,24 +79,228 @@ mod tests {
+        ///
+        /// This method will panic if there are not two profiles.
+        fn query(test_name: &str) -> (RawProfile, RawProfile) {
+            let pattern = format!("target/**/*{}.raw", test_name);
+            let raw_files: Vec<String> = glob::glob(&pattern)
+                .expect("Failed to read glob pattern")
+                .filter_map(Result::ok)
+                .map(|path| path.to_string_lossy().into_owned())
+                .collect();
+            assert_eq!(raw_files.len(), 2);
+
+            let profile1 = RawProfile {
+                test_name: test_name.to_string(),
+                commit_hash: git::extract_commit_hash(&raw_files[0]),
+            };
+
+            let profile2 = RawProfile {
+                test_name: test_name.to_string(),
+                commit_hash: git::extract_commit_hash(&raw_files[1]),
+            };
+
+            assert!(git::is_commit_in_log(&profile1.commit_hash));
+            assert!(git::is_commit_in_log(&profile2.commit_hash));
+
+            if git::is_older_commit(&profile1.commit_hash, &profile2.commit_hash) {
+                (profile1, profile2)
+            } else {
+                (profile2, profile1)
+            }
+        }
+    }
+
+    struct AnnotatedProfile {
+        test_name: String,
+        commit_hash: String,
+    }
+
+    impl AnnotatedProfile {
+        fn new(raw_profile: &RawProfile) -> Self {
+            let annotated = Self {
+                test_name: raw_profile.test_name.clone(),
+                commit_hash: raw_profile.commit_hash.clone(),
+            };
+
+            // annotate raw profile
+            let annotate_output = Command::new("cg_annotate")
+                .arg(raw_profile.path())
+                .output()
+                .expect("Failed to run cg_annotate");
+            assert!(annotate_output.status.success());
+
+            // write annotated profile to disk
+            let annotate_content = String::from_utf8(annotate_output.stdout)
+                .expect("Invalid UTF-8 in cg_annotate output");
+            write(annotated.path(), annotate_content).expect("Failed to write to file");
+
+            annotated
+        }
+
+        fn path(&self) -> String {
+            format!("target/{}/{}.annotated", self.commit_hash, self.test_name)
         }
     }
 
+    struct DiffProfile {
+        test_name: String,
+    }
+    impl DiffProfile {
+        fn new(prev_profile: &RawProfile, curr_profile: &RawProfile) -> Self {
+            let diff_profile = Self {
+                test_name: curr_profile.test_name.clone(),
+            };
+
+            // diff the raw profile
+            let diff_output = Command::new("cg_annotate")
+                .args(["--diff", &prev_profile.path(), &curr_profile.path()])
+                .output()
+                .expect("Failed to run cg_annotate --diff");
+            assert!(diff_output.status.success());
```

Comment:
this might be difficult to debug if you don't print out `stdout`/`stderr`. Same goes for all of the other commands.